core/modules/replica/IngestRequest.cc

```diff
@@ -0,0 +1,600 @@
+                + TransactionContribInfo::status2str(TransactionContribInfo::Status::IN_PROGRESS)
+                + ", the actual state is " + TransactionContribInfo::status2str(contrib.status) + ".");
+    }
+    if (!contrib.async) {
+        throw invalid_argument("contribution id=" + to_string(contribId) + " is not ASYNC.");
+    }
+
+    // Note that contrib.startTime doesn't need to be validated since it's allowed
+    // to resume requests that have not been started yet or which are still in an early
+    // processing state (before the final stage when changes to MySQL are about to be
+    // made or have been made).
+    if ((contrib.createTime == 0) || (contrib.readTime != 0) || (contrib.loadTime != 0)) {
+        throw invalid_argument(
+                "contribution id=" + to_string(contribId) + " is not eligible to be resumed since"
+                " changes to the MySQL table may have already been made.");
+    }
+
+    auto const trans = databaseServices->transaction(contrib.transactionId);
+    if (trans.state != TransactionInfo::STARTED) {
+        contrib.status = TransactionContribInfo::Status::CREATE_FAILED;
+        contrib.error = context + " transactionId=" + to_string(contrib.transactionId) + " is not active";
+        contrib = databaseServices->updateTransactionContrib(contrib);
+        throw invalid_argument(contrib.error);
+    }
+
+    // Make sure the state is clear (except the contrib.id and contrib.createTime
+    // which need to be retained)
+    contrib.startTime = 0;
+    contrib.tmpFile = string();
+    contrib.error = string();
+    contrib.httpError = 0;
+    contrib.systemError = 0;
+    contrib.retryAllowed = false;
+    contrib = databaseServices->updateTransactionContrib(contrib);
+
+    return IngestRequest::Ptr(new IngestRequest(serviceProvider, workerName, contrib));
+}
+
+
+IngestRequest::IngestRequest(
+        ServiceProvider::Ptr const& serviceProvider,
+        string const& workerName,
+        TransactionId transactionId,
+        string const& table,
+        unsigned int chunk,
+        bool isOverlap,
+        string const& url,
+        bool async,
+        string const& fieldsTerminatedBy,
+        string const& fieldsEnclosedBy,
+        string const& fieldsEscapedBy,
+        string const& linesTerminatedBy,
+        string const& httpMethod,
+        string const& httpData,
+        vector<string> const& httpHeaders)
+        :   IngestFileSvc(serviceProvider, workerName) {
+
+    // Initialize the descriptor
+    _contrib.transactionId = transactionId;
+    _contrib.table = table;
+    _contrib.chunk = chunk;
+    _contrib.isOverlap = isOverlap;
+    _contrib.worker = workerInfo().name;
+    _contrib.url = url;
+    _contrib.async = async;
+    _contrib.fieldsTerminatedBy = fieldsTerminatedBy;
+    _contrib.fieldsEnclosedBy = fieldsEnclosedBy;
+    _contrib.fieldsEscapedBy = fieldsEscapedBy;
+    _contrib.linesTerminatedBy = linesTerminatedBy;
```

Comment:
These four things are always used together, can you make a simple struct for them, could save you some typing.